### ISSUE TITLE
backbone.select incompatible with Backbone.Collection initialize with {parse:true}

### ISSUE THREAD

**Issue by nmackey** [hashchange/backbone.select#11]
Although not really documented (I had to go to the source code to verify) a Backbone.Collection will let you initialize a collection with an object that may not be an array of models with the ```{parse:true}``` option.  Backbone will then call the parse function on the object passed in, which can convert the object passed in to an array of models.  However backbone.select requires the 'models' passed in to be an array.

I noticed this in tests where I setup my collections with a json object that mimics my server call and then I pass that to my new collection with ```{parse: true}```.  This worked fine with backbone.picky (since the models weren't passed in to it), however it breaks backbone.select

I am able to work around it since it is in my test code, however it does seem to be an incompatibility with the Backbone API on backbone.select's part.

**Comment by hashchange** [comment]
Interesting catch ... and thank you for your helpful description! I'll deal with the issue further down the road. I can't say for sure that I'll actually fix it - after all, the issue relates to undocumented Backbone behaviour. On the other hand, Backbone is meant to be hacked, so it would make sense to make it work. I'll have a look.

**Comment by hashchange** [comment]
Sorry for the delay ... I've had a closer look now. As far as I can see, a fix would have to replace Backbone.Collection with a modified implementation. 

The issue is that Backbone parses the models and creates the `models` property after `initialize()` is called. That order would have to be reversed. The fix I have in mind would look something like this:

```javascript
var originalConstructor = Backbone.Collection;

Backbone.Collection = Backbone.Collection.extend( {

    constructor: function ( models, options ) {

        if ( options && options.parse && !this._isModel( models ) ) { 
            models = this.parse( models, options ) || [];
            options = _.extend( {}, options, { parse: false } );
        }

        originalConstructor.call( this, models, options );

    }

} );
```

That would do the trick (presumably), but I don't like it for two reasons. 

First, the options object handed to `initialize()` is no longer the one which was passed in. `initialize()` receives a clone, and the `parse` option has changed to `false`, in order to prevent a second parse call. That alone might be enough to break stuff for some people.

Second, it's a heavy-handed approach. It would place otherwise unneccessary constraints on the load order of components: Backbone.Select would have to be loaded after Backbone, but before anything else.

Imagine the fallout e.g. in an AMD setup. Every component which depends on Backbone would have to list Backbone.Select as a dependency, too. A shim would be required for pretty much every Backbone-related third-party module. I don't think that's worth it, given that we are dealing with an undocumented feature. 

So I'll leave things as they are. However, if you disagree or have a different idea for a fix, just let me know. And thanks for the report!

**Comment by hashchange** [comment]
I retract everything I've said ;) Seems like I'll be able to fix it cleanly in the upcoming version 2.

**Comment by nmackey** [comment]
Sounds great, thanks.

**Comment by hashchange** [comment]
Has landed in 2.0.0.
